src/app/modules/service-providers/services/navigation/navigation.service.ts

```diff
@@ -1,16 +1,101 @@
 import { Injectable } from '@angular/core';
-import { Store } from '@ngrx/store';
+import { select, Store } from '@ngrx/store';
+import { lineSlice, lineString, nearestPointOnLine, point, length } from '@turf/turf';
+import {
+  currentPositionSelector, getNavState,
+  nextWaypointIndexSelector,
+  routeDetailsSelector,
+  routePointsSelector
+} from '../../../../store/nav.selectors';
+import { filter, tap, withLatestFrom } from 'rxjs/operators';
+import {
+  PlayVoiceWpNotificationAction,
+  SetDistanceToEndpointAction,
+  SetIsNextWpNotifiedAction,
+  SetNextWaypointDistanceAction,
+  SetNextWaypointIndexAction
+} from '../../../../store/nav.actions';
+
 
 @Injectable({
   providedIn: 'root'
 })
 export class NavigationService {
 
-  constructor(private store: Store<any>) { }
+  constructor(private store: Store<any>) {
+    this.calcNextWaypoints();
+  }
+
+
+  calcNextWaypoints() {
+    this.store.pipe(
+      select(currentPositionSelector),
+      withLatestFrom(this.store.pipe(select(routePointsSelector))),
+      filter(([currentPosition, routePoints]) => currentPosition.latitude !== null && routePoints !== null),
+      withLatestFrom(this.store.pipe(select(nextWaypointIndexSelector)), this.store.pipe(select(routeDetailsSelector))),
+      tap(([[currentPosition, routePoints], nextWpIndex, routeDetails]) => {
+        const currentPositionOnRoute = nearestPointOnLine(lineString(routePoints), point([currentPosition.latitude,
```

Comment:
Refactor to multiple methods for readability, current code is too long.
updateDistanceToEndpoint
checkAndUpdateNextWaypoint
updateDistanceToNextWaypoint